teps/0021-results-api.md

```diff
@@ -0,0 +1,772 @@
+Tekton's API surface being limited to custom resource objects stored in a
+cluster's etcd also means that execution information not described as a CRD get
+dropped entirely. For example, triggering invocations (event received, k8s
+object created) are not stored in etcd, and so there's no way at all to query
+them.
+
+### Querying
+
+Being implemented as a set of Kubernetes custom resource objects stored in etcd,
+users can query the Kubernetes API server for the state of a single resource by
+name easily and efficiently. However, Kubernetes' support for listing/querying
+across many custom resources is severely lacking.
+
+Kubernetes allows users to request filtered lists of API resources using
+[field selectors and label selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/).
+For custom resources the only support field selectors are name and namespace.
+Label selectors are likewise limited, since we'd have to store any queryable
+information about a Run in a label for the Run to be able to query for it, and
+this grows large quickly. Together, these limitations make it pretty much
+impossible to query for anything actually useful today.
+
+### Goals
+
+<!--
+List the specific goals of the TEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+- Define a Result API spec to store and query Tekton results.
+  - Tekton results are any data about or related to TaskRuns or PipelineRuns. At
+    minimum this includes the full TaskRun and PipelineRun status/spec, but may
+    include other data about the Runs (events that triggered it, artifacts
+    produced, post-run tasks, etc.)
+- Decouple long-term Tekton result storage away from runtime storage in etcd.
+- Give clients flexibility to embed additional metadata associated with Tekton
+  Task/PipelineRuns.
+
+### Non-Goals
+
+<!--
+What is out of scope for this TEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Log exporting - we will rely on
+  [existing projects](https://landscape.cncf.io/category=logging&format=card-mode&grouping=category)
+  to handle exporting of logs to remote systems.
+- Defining a system to automatically clean up completed Task/PipelineRuns.
+- Define specific authentatication/authorization mechanisms (this will likely be
+  implementation specific).
+
+## Requirements
+
+<!--
+List the requirements for this TEP.
+-->
+
+- Results integrations should be an optional add-on on top of Tekton. Tekton
+  Pipelines should be able to operate normally without the presense of a Result
+  server.
+- Result API Server hosts should be able to store multi-tenant results and share
+  the server among multiple namespaces or clusters.
+- Service providers can choose to provide their own controller and server
+  implementations to upload Results with additional service specific metadata or
+  implementation specific validation / field formats, so long as it conforms to
+  the Results API.
```

Comment:
great requirements 👍  i think these can be good examples to point to if folks are trying to distinguish b/w goals + requirements in other TEPs